Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,203 @@
+
+Then the Loader will pass to the decoder the key-value pair object of state that is required to recreate the object and restore it's state. The best way to show this even though it will largely be left up to interface implementers to do this, is with an example:
+```python
+class exampleDecoder():
+  def decode(self, dict):
+    testui = TestUI(dict["arg1"], dict["args2"])
+    testui.ui.spinBox.setValue(dict["spinBox"]) 
+    testui.ui.lineEdit.setValue(dict["lineEdit"]) 
+    testui.ui.dateEdit.setValue(dict["dateEdit"]) 
+    testui.ui.radioButton.setValue(dict["radioButton"]) 
+    testui.ui.checkBox.setValue(dict["checkBox"])
+    testui.ui.label_7.setVisible(dict["label_7Visible"])
+    return testui
+```
+
+Then this leaves custom Decoding from JSON to be tackled, a custom decoder is a little harder, than an encoder, as it relies on the key name when decoding being something unique to that type so would need to be on a type requirement by requirement basis. So in the JSON a programmer would need to add a _object_type_ or some other unique string to show that a type like that is there, but it's really on a per type basis. For example: 
+```python
+# (Docs.python.org, 2018)
+def object_type(dict):
+  # __object_type__ being a key that is unique to that type
+  if "__object_type__" in dict:
+    # Return an object of the original object type
+    return object(dict["object_part1"], dict["object_part2"], dict["object_part3"])
+  return dict
+```
+
+**C++ Saving and Loading Integration:**
+Similarly with the C++ area of serialisation exposing most of the required details to python will be the hardest part, however if most of the interface's details are exposed properly then the decoder and encoders should be able to work exactly the same on these interfaces if only with a little extra implementation on the UI porter/maintainer's part.
+
+Implementation of save workspaces and linking to saved scripts
+--------------------------------------------------------------
+Utilizing the already implemented SaveNexusProcessed and SaveMD algorithms, the suggestion is to implement a python version of how MantidPlot currently implements saving workspaces. Call SaveNexusProcessed for most workspaces and SaveMD for EventWorkspaces. Saving a copy of the names of the saved workspace .nxs files in a JSON format in the same file as the output from the serialization of the windows, it may look like this:
```

Comment:
Please note that `SaveNexusProcessed` does not work on `MaskWorkspace`s and thus our current project saving does not save/restore such workspaces. One needs to use `SaveMask` instead.

Reply:
Noted and will update this document accordingly. The caveat for changing workspace type is already in place of EventWorkspaces so it would just be adding another one to list :+1: